File: objective-c/Database/library/RunParser.m
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#import <objc/Ice.h>
#import <objc/Glacier2.h>
#import <Library.h>
#import <Session.h>
#import <Glacier2Session.h>
#import <Parser.h>
#import <stdio.h>
#import <string.h>

int
runParser(int argc, char* argv[], id<ICECommunicator> communicator)
{
    id<GLACIER2RouterPrx> router = [GLACIER2RouterPrx uncheckedCast:[communicator getDefaultRouter]];
    id<DemoLibraryPrx> library;
    id session;
    if(router != nil)
    {
        printf("This demo accepts any user-id / password combination.\n");

        while(true)
        {
            char id[1024];
            printf("user id: ");
            fflush(stdout);
            fgets(id, sizeof(id), stdin);
            int len = strlen(id);
            if(id[len-1] == '\n')
            {
                id[len-1] = '\0';
            }
            printf("password: ");
            fflush(stdout);
            char pw[1024];
            fgets(pw, sizeof(pw), stdin);
            len = strlen(pw);
            if(pw[len-1] == '\n')
            {
                pw[len-1] = '\0';
            }

            @try
            {
                session = [DemoGlacier2SessionPrx uncheckedCast:
                            [router createSession:[NSString stringWithCString:id encoding:NSUTF8StringEncoding]
                            password:[NSString stringWithCString:pw encoding:NSUTF8StringEncoding]]];
                break;
            }
            @catch(GLACIER2PermissionDeniedException* ex)
            {
                printf("permission denied:\n%s\n", [ex.reason UTF8String]);
            }
            @catch(GLACIER2CannotCreateSessionException* ex)
            {
                printf("cannot create session:\n%s\n", [ex.reason UTF8String]);
            }
        }
        library = [session getLibrary];

        ICELong acmTimeout = [router getACMTimeout];
        if(acmTimeout > 0)
        {
            //
            // Configure the connection to send heartbeats in order to keep our session alive
            //
            [[router ice_getCachedConnection] setACM:@(acmTimeout) close:ICENone heartbeat:@(ICEHeartbeatAlways)];
        }
    }
    else
    {
        id<DemoSessionFactoryPrx> factory = [DemoSessionFactoryPrx checkedCast:[
                communicator propertyToProxy:@"SessionFactory.Proxy"] ];
        if(factory == nil)
        {
            fprintf(stderr, "%s: invalid object reference", argv[0]);
            return 1;
        }

        session = [factory create];
        library = [session getLibrary];
    }

    int rc = [[Parser parserWithLibrary:library] parse];

    if(router)
    {
        @try
        {
            [router destroySession];
        }
        @catch(GLACIER2SessionNotExistException* ex)
        {
            NSLog(@"%@\n", [ex description]);
        }
        @catch(ICEConnectionLostException* ex)
        {
            //
            // Expected: the router closed the connection.
            //
        }
    }
    else
    {
        [session destroy];
    }

    return rc;
}

File: java/Database/library/RunParser.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import com.zeroc.demos.Database.library.Demo.*;

import com.zeroc.Ice.ACMClose;
import com.zeroc.Ice.ACMHeartbeat;

class RunParser
{
    //
    // Adapter for the two types of session objects.
    //
    interface SessionAdapter
    {
        public LibraryPrx getLibrary();
        public void destroy();
    }

    static SessionAdapter createSession(String appName, com.zeroc.Ice.Communicator communicator)
    {
        SessionAdapter session;
        final com.zeroc.Glacier2.RouterPrx router =
            com.zeroc.Glacier2.RouterPrx.uncheckedCast(communicator.getDefaultRouter());
        if(router != null)
        {
            com.zeroc.Glacier2.SessionPrx glacier2session = null;
            long timeout;
            java.io.BufferedReader in = new java.io.BufferedReader(new java.io.InputStreamReader(System.in));
            while(true)
            {
                System.out.println("This demo accepts any user-id / password combination.");
                try
                {
                    String id;
                    System.out.print("user id: ");
                    System.out.flush();
                    id = in.readLine();

                    String pw;
                    System.out.print("password: ");
                    System.out.flush();
                    pw = in.readLine();

                    try
                    {
                        glacier2session = router.createSession(id, pw);
                        final int acmTimeout = router.getACMTimeout();
                        if(acmTimeout > 0)
                        {
                            com.zeroc.Ice.Connection connection = router.ice_getCachedConnection();
                            assert(connection != null);
                            connection.setACM(
                                java.util.OptionalInt.of(acmTimeout), null,
                                java.util.Optional.of(ACMHeartbeat.HeartbeatAlways));
                        }
                        break;
                    }
                    catch(com.zeroc.Glacier2.PermissionDeniedException ex)
                    {
                        System.out.println("permission denied:\n" + ex.reason);
                    }
                    catch(com.zeroc.Glacier2.CannotCreateSessionException ex)
                    {
                        System.out.println("cannot create session:\n" + ex.reason);
                    }
                }
                catch(java.io.IOException ex)
                {
                    ex.printStackTrace();
                }
            }

            final Glacier2SessionPrx sess = Glacier2SessionPrx.uncheckedCast(glacier2session);
            session = new SessionAdapter()
            {
                @Override
                public LibraryPrx getLibrary()
                {
                    return sess.getLibrary();
                }

                @Override
                public void destroy()
                {
                    try
                    {
                        router.destroySession();
                    }
                    catch(com.zeroc.Glacier2.SessionNotExistException ex)
                    {
                    }
                    catch(com.zeroc.Ice.ConnectionLostException ex)
                    {
                        //
                        // Expected: the router closed the connection.
                        //
                    }
                }
            };
        }
        else
        {
            SessionFactoryPrx factory = SessionFactoryPrx.checkedCast(
                communicator.propertyToProxy("SessionFactory.Proxy"));
            if(factory == null)
            {
                System.err.println(appName + ": invalid object reference");
                return null;
            }

            final SessionPrx sess = factory.create();
            session = new SessionAdapter()
            {
                @Override
                public LibraryPrx getLibrary()
                {
                    return sess.getLibrary();
                }

                @Override
                public void destroy()
                {
                    sess.destroy();
                }
            };
        }
        return session;
    }

    static int runParser(String appName, String[] args, final com.zeroc.Ice.Communicator communicator)
    {
        final SessionAdapter session = createSession(appName, communicator);
        if(session == null)
        {
            return 1;
        }

        LibraryPrx library = session.getLibrary();

        Parser parser = new Parser(communicator, library);

        int rc = 0;

        if(args.length == 1)
        {
            rc = parser.parse(args[0]);
        }

        if(rc == 0)
        {
            rc = parser.parse();
        }

        session.destroy();

        return rc;
    }
}
